### PR TITLE
feat: allow params number

### PR DESCRIPTION
**Pull request by pikax**
Allow have `param` with number as a value

### CODE REVIEW

File: __tests__/router.spec.ts [hunk 1]
```diff
@@ -253,6 +253,16 @@ describe('Router', () => {
     expect(router.currentRoute.value.hash).toBe('#two')
   })
 
+  it('allow param value to be number', async () => {
```
Comment (posva) [inline __tests__/router.spec.ts:256]:
```suggestion
  it('cast number params to string', async () => {
```

File: __tests__/router.spec.ts [hunk 2]
```diff
@@ -253,6 +253,16 @@ describe('Router', () => {
...
+    const { router } = await newRouter()
+    router.addRoute({ path: '/r1/:r', name: 'r1', component: components.Bar })
+    expect(router.resolve({ name: 'r1', params: { r: 1 } })).toMatchObject({
+      name: 'r1',
+      path: '/r1/1',
+      params: { r: '1' },
```
Comment (posva) [inline __tests__/router.spec.ts:262]: let's not add a new route and use the one named Param
```suggestion
    expect(router.resolve({ name: 'Param', params: { p: 1 } })).toMatchObject({
      name: 'Param',
      path: '/p/1',
      params: { p: '1' },
```

File: src/router.ts [hunk 3]
```diff
@@ -264,7 +264,10 @@ export function createRouter(options: RouterOptions): Router {
       // for same reason TS thinks rawLocation.params can be undefined
       matcher.resolve(
         'params' in rawLocation
-          ? { ...rawLocation, params: encodeParams(rawLocation.params) }
+          ? {
+              ...rawLocation,
+              params: encodeParams(applyToParamsRaw(rawLocation.params)),
```
Comment (posva) [inline src/router.ts:269]: A few thoughts for refactoring:

- `encodeParams` could accept numbers
- we could create a copy of `rawLocation` before L256, and add an else at L261 to change the `params` property to an encoded version without numbers, save the decoded version as a local variable `params` to use it later

File: src/utils/index.ts [hunk 4]
```diff
@@ -20,3 +36,9 @@ export function applyToParams(
 
   return newParams
 }
+
+export function applyToParamsRaw(
+  params: RouteParamsRaw | undefined
+): RouteParams | undefined {
+  return applyToParams(x => x.toString(), params)
+}
```
Comment (posva) [inline src/utils/index.ts:44]: I think we could write this one as

```suggestion
export let normalizeParams = applyParams.bind(null, paramValue => '' + paramValue)
```

File: src/router.ts [hunk 5]
```diff
@@ -275,7 +278,7 @@ export function createRouter(options: RouterOptions): Router {
     // TODO: normalize params if we accept numbers as raw values
     matchedRoute.params =
       'params' in rawLocation
-        ? rawLocation.params!
+        ? applyToParamsRaw(rawLocation.params)!
```
Comment (posva) [inline src/router.ts:281]: we can remove the todo above

File: src/utils/index.ts (hunk 4 above)
Comment (pikax) [reply src/utils/index.ts:44]: noticed `let` instead of `const`, is that on purpose? If it is, why? :)

File: src/router.ts (hunk 3 above)
Comment (pikax) [reply src/router.ts:269]: `encodeParams` is accepting `string | number` but the `bind` doesn't pick up correct overload, I will just create a const with the `normalized` params and use that across

**Comment by posva** [comment]
I ended up doing something simpler by not allowing number params for the matcher at ef0920a86574bca10836214015c2317ed11a29b7
I still kept you as the author of the commit, I hope that's fine

**Pull request closed without merging**